### PR TITLE
[MODINREACH-122] INN-Reach Circulation Flow: React to Change in Request Record for INN-Reach Transaction: Transfer (Move) Request (Item Hold)

### PR DESCRIPTION
**Pull request by aleksandr2110**
## Purpose
https://issues.folio.org/browse/MODINREACH-122

## Approach
mod-inn-reach needs to be aware of and respond (when necessary) to moving a request from one item to another. The module should consume appropriate domain events for requests and respond accordingly when indicated.
1. mod-inn-reach should consume request-storage domain events via Kafka ([CIRCSTORE-302](https://issues.folio.org/browse/CIRCSTORE-302))
2. mod-inn-reach should check if the request is associated with an active INN-Reach Item Hold transaction
3. If so, and the item associated with the request does not match the item currently associated with the INN-Reach transaction, update the referenced item record in the INN-Reach transaction and issue a transfer request message to the central server (D2IR API Reference pg. 64) and update the transaction status to TRANSFER

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java [hunk 1]
```diff
@@ -169,6 +172,22 @@ public void handleLoanUpdate(LoanDTO loan) {
...
+    if (transaction == null) {
+      return;
+    }
+
+    TransactionHold transactionItemHold = transaction.getHold();
+    if (!transactionItemHold.getFolioItemId().equals(requestDTO.getItemId())) {
+      log.info("Updating transaction {} on request to transfer {}", transaction.getId(),requestDTO.getId());
+      transactionItemHold.setFolioItemId(requestDTO.getItemId());
+      transaction.setState(TRANSFER);
+      reportTransferRequest(transaction, requestDTO.getItemId().toString());
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: we link items by hrid when contributing them to central server, so I assume here it also should be referenced by hrid. Ask PO about this. 
Also, check the rules for FOLIO to D2IR fields mapping https://wiki.folio.org/display/FOLIJET/SPIKE%3A+%5BMODINREACH-79%5D+Record+Contribution%3A+Analyze+data+required+for+D2IR+contribution

File: src/main/java/org/folio/innreach/domain/entity/InnReachTransaction.java [hunk 2]
```diff
@@ -100,6 +106,11 @@
     "JOIN FETCH h.pickupLocation " +
     "WHERE h.folioLoanId = :folioLoanId AND irt.state NOT IN (4, 11, 12, 13)";
 
+  public static final String FETCH_BY_REQUEST_ID_QUERY_NAME = "InnReachTransaction.fetchTransactionItemHoldByHoldingsRecord";
+  public static final String FETCH_BY_REQUEST_ID_QUERY = "SELECT irt FROM InnReachTransaction AS irt " +
+    "JOIN FETCH irt.hold AS h " +
+    "JOIN FETCH h.pickupLocation " + "WHERE h.folioRequestId = :folioRequestId AND irt.state NOT IN (4, 11, 12, 13)";
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/entity/InnReachTransaction.java]: reuse FETCH_OPEN_TRANSACTIONS with appending condition for request id check

File: src/main/java/org/folio/innreach/domain/listener/KafkaCirculationEventListener.java [hunk 3]
```diff
@@ -52,6 +53,33 @@ public void handleLoanEvents(List<ConsumerRecord<String, DomainEvent<LoanDTO>>>
...
+      .collect(Collectors.toList());
+
+    eventProcessor.process(events, event -> {
+      switch (event.getType()) {
+      case CREATED:
+        //
+        break;
+      case UPDATED:
+        transactionActionService.handleRequestUpdate(event.getData().getNewEntity());
+        // request is associated with an active INN-Reach Item Hold transaction
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/listener/KafkaCirculationEventListener.java]: missing `break;`

Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/listener/KafkaCirculationEventListener.java]: it's actually unknown at this point if request is associated with any transaction. Remove this comment or move it to the service method

File: src/main/java/org/folio/innreach/repository/InnReachTransactionRepository.java [hunk 4]
```diff
@@ -34,4 +35,6 @@
   @Query(name = FETCH_OPEN_BY_LOAN_ID_QUERY_NAME)
   Optional<InnReachTransaction> fetchOneByLoanId(UUID folioLoanId);
 
+  @Query(name = FETCH_BY_REQUEST_ID_QUERY_NAME)
+  Optional<InnReachTransaction> fetchOneByRequestId(UUID folioRequestId);
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/repository/InnReachTransactionRepository.java]: fetchActiveByRequestId/fetchOpenByRequestId

Comment (oliinyko) [reply src/main/java/org/folio/innreach/repository/InnReachTransactionRepository.java]: same for the query name

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java [hunk 5]
```diff
@@ -169,6 +172,22 @@ public void handleLoanUpdate(LoanDTO loan) {
...
   }
 
+  @Override
+  public void handleRequestUpdate(RequestDTO requestDTO) {
+    var transaction = transactionRepository.fetchOneByRequestId(requestDTO.getId()).orElse(null);
+    if (transaction == null) {
+      return;
+    }
+
+    TransactionHold transactionItemHold = transaction.getHold();
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: var?

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java [hunk 6]
```diff
@@ -169,6 +172,22 @@ public void handleLoanUpdate(LoanDTO loan) {
...
+    if (!transactionItemHold.getFolioItemId().equals(requestDTO.getItemId())) {
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java:189]: > mod-inn-reach should check if the request is associated with an active INN-Reach **Item Hold** transaction

File: src/main/java/org/folio/innreach/domain/listener/KafkaCirculationEventListener.java (hunk 3 above)
Comment (aleksandr2110) [reply src/main/java/org/folio/innreach/domain/listener/KafkaCirculationEventListener.java]: Done

Comment (aleksandr2110) [reply src/main/java/org/folio/innreach/domain/listener/KafkaCirculationEventListener.java]: Done

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java (hunk 5 above)
Comment (aleksandr2110) [reply src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: Fixed

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java (hunk 6 above)
Comment (aleksandr2110) [reply src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java:189]: We check it here  var transaction = transactionRepository.fetchOneByRequestId(requestDTO.getId()).orElse(null);

Comment (oliinyko) [reply src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java:189]: it doesnt check if transaction is Item Hold

File: src/main/java/org/folio/innreach/repository/InnReachTransactionRepository.java (hunk 4 above)
Comment (aleksandr2110) [reply src/main/java/org/folio/innreach/repository/InnReachTransactionRepository.java]: Fixed

File: src/main/java/org/folio/innreach/domain/entity/InnReachTransaction.java (hunk 2 above)
Comment (aleksandr2110) [reply src/main/java/org/folio/innreach/domain/entity/InnReachTransaction.java]: Fixed

File: src/main/java/org/folio/innreach/domain/listener/KafkaCirculationEventListener.java (hunk 3 above)
Comment (aleksandr2110) [reply src/main/java/org/folio/innreach/domain/listener/KafkaCirculationEventListener.java]: Deleted

Comment (aleksandr2110) [reply src/main/java/org/folio/innreach/domain/listener/KafkaCirculationEventListener.java:76]: Added

File: src/main/java/org/folio/innreach/domain/entity/InnReachTransaction.java (hunk 2 above)
Comment (aleksandr2110) [reply src/main/java/org/folio/innreach/domain/entity/InnReachTransaction.java:112]: Fixed

File: src/main/java/org/folio/innreach/repository/InnReachTransactionRepository.java (hunk 4 above)
Comment (aleksandr2110) [reply src/main/java/org/folio/innreach/repository/InnReachTransactionRepository.java]: Fixed

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java (hunk 5 above)
Comment (aleksandr2110) [reply src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java:182]: Chanced

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java (hunk 1 above)
Comment (aleksandr2110) [reply src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: Fixed

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java (hunk 6 above)
Comment (aleksandr2110) [reply src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java:189]: Fixed

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java [hunk 7]
```diff
@@ -169,6 +174,29 @@ public void handleLoanUpdate(LoanDTO loan) {
...
+
+    if (transaction.getType() != ITEM) {
+      log.info("Transaction {} isn't Item Hold", transaction.getId());
+      return;
+    }
+
+    var transactionItemHold = (TransactionItemHold) transaction.getHold();
+    if (!transactionItemHold.getFolioItemId().equals(requestDTO.getItemId())) {
+      log.info("Updating transaction {} on request to transfer {}", transaction.getId(), requestDTO.getId());
+      transactionItemHold.setFolioItemId(requestDTO.getItemId());
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java:191]: there are multiple fields that potentially need to be updated when the request is moved to another item:
```
  folioInstanceId
  folioHoldingId
  folioItemId
  folioItemBarcode
```

Comment (aleksandr2110) [reply src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java:191]: Resolved

File: src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java [hunk 8]
```diff
@@ -232,6 +242,88 @@ void shouldUpdateTransactionToCheckIn() {
...
+    var capturedEvents = eventsCaptor.getValue();
+    assertEquals(1, capturedEvents.size());
+
+    var capturedEvent = capturedEvents.get(0);
+    assertEquals(folioRequestId, capturedEvent.getRecordId());
+    var transaction = transactionRepository.fetchActiveByRequestId(requestDTO.getId()).orElse(null);
+    var transactionItemHold = (TransactionItemHold) transaction.getHold();
+
+    assertEquals(ITEM_ID, transactionItemHold.getFolioItemId());
+    assertEquals(INSTANCE_ID, transactionItemHold.getFolioInstanceId());
```
Comment (oliinyko) [inline src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java:280]: add asserts for holding id and item barcode

File: src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java [hunk 9]
```diff
@@ -232,6 +242,88 @@ void shouldUpdateTransactionToCheckIn() {
...
+    var consumerRecord = new ConsumerRecord(CIRC_REQUEST_TOPIC, 1, 1, folioRequestId.toString(), event);
+    listener.handleRequestStorage(List.of(consumerRecord));
+
+    ArgumentCaptor<List<DomainEvent<RequestDTO>>> eventsCaptor = ArgumentCaptor.forClass(List.class);
+    verify(eventProcessor).process(eventsCaptor.capture(), any(Consumer.class));
+    var capturedEvents = eventsCaptor.getValue();
+
+    var capturedEvent = capturedEvents.get(0);
+    assertEquals(folioRequestId, capturedEvent.getRecordId());
+  }
```
Comment (oliinyko) [inline src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java]: how does this test verifies that the event is "skipped" ?

File: src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java (hunk 8 above)
Comment (oliinyko) [reply src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java:280]: add assert to verify that "D2IR transfer request" is called

Comment (aleksandr2110) [reply src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java:280]: Fixed

File: src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java (hunk 9 above)
Comment (aleksandr2110) [reply src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java]: Fixed

File: src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java [hunk 10]
```diff
@@ -232,6 +244,95 @@ void shouldUpdateTransactionToCheckIn() {
...
+
+    ArgumentCaptor<List<DomainEvent<RequestDTO>>> eventsCaptor = ArgumentCaptor.forClass(List.class);
+    verify(eventProcessor).process(eventsCaptor.capture(), any(Consumer.class));
+    var capturedEvents = eventsCaptor.getValue();
+
+    var capturedEvent = capturedEvents.get(0);
+    assertEquals(folioRequestId, capturedEvent.getRecordId());
+    verify(transactionRepository, times(1)).fetchActiveByRequestId(any());
+    verify(itemService, times(0)).find(any());
+    verify(innReachExternalService, times(0)).postInnReachApi(any(), any(), any());
```
Comment (oliinyko) [inline src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java:333]: just a tip, there is `verifyNoInteractions` for this purpose

**Review by oliinyko — APPROVED** [review]

File: src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java (hunk 10 above)
Comment (aleksandr2110) [reply src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java:333]: It is new to me. Thank you.

**Pull request merged**